subt/docker/unittest/subt_seed_node.cc

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (C) 2018 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+// UnitTest attempt based on subt_seed_node.cc
+
+#include <chrono>
+#include <geometry_msgs/Twist.h>
+#include <rosgraph_msgs/Clock.h>
+#include <sensor_msgs/Imu.h>
+#include <sensor_msgs/LaserScan.h>
+#include <sensor_msgs/CompressedImage.h>
+#include <nav_msgs/Odometry.h>
+
+#include <subt_msgs/PoseFromArtifact.h>
+#include <ros/ros.h>
+#include <std_srvs/SetBool.h>
+#include <rosgraph_msgs/Clock.h>
+
+#include <string>
+#include <stdlib.h>     /* abs */
+
+#include <subt_communication_broker/subt_communication_client.h>
+#include <subt_ign/CommonTypes.hh>
+#include <subt_ign/protobuf/artifact.pb.h>
+
+int g_countClock = 0;
+int g_countImu = 0;
+int g_countScan = 0;
+int g_countImage = 0;
+int g_countOdom = 0;
+
+
+void clockCallback(const rosgraph_msgs::Clock::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  if(g_countClock % 1000 == 0)
+    ROS_INFO("received Clock %d ", g_countClock);
+  g_countClock++;
+}
+
+void imuCallback(const sensor_msgs::Imu::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  if(g_countImu % 100 == 0)
+    ROS_INFO("received Imu %d ", g_countImu);
+  g_countImu++;
+}
+
+void scanCallback(const sensor_msgs::LaserScan::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  if(g_countScan % 100 == 0)
+    ROS_INFO("received Scan %d", g_countScan);
+  g_countScan++;
+}
+
+void imageCallback(const sensor_msgs::CompressedImage::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  if(g_countImage % 100 == 0)
+    ROS_INFO("received Image %d", g_countImage);
+  g_countImage++;
+}
+
+void odomCallback(const nav_msgs::Odometry::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  if(g_countOdom % 100 == 0)
+    ROS_INFO("received Odom %d", g_countOdom);
+  g_countOdom++;
```

Comment:
Since there is the [header field](http://wiki.ros.org/Messages), it might be beneficial to check the `seq` field is increasing by 1 and that the `stamp` has some meaningful relation to the value of `/clock`. Just a suggestion.

Reply:
Good idea :). I would maybe first merge it and then revise it from master in new branch? I am still thinking if to check `/robot_data` option how to save data via ROS bag (which is must have for our solution). Thanks.

Reply:
Whatever works for you. At the end, both should be in the unittest since without it, there is no point in participating.